sudoku_board.h

```diff
@@ -1,40 +1,107 @@
-    std::vector<std::vector<int>> m_board_values;
-    std::vector<std::vector<std::set<int>>> m_possible_options;
-
-    std::set<int> get_values_in_cube(int r, int c);
-    std::set<int> get_values_in_row(int r, int c);
-    std::set<int> get_values_in_col(int r, int c);
-    std::set<int> get_forbidden_values(int r, int c);
-    void empty_options();
+
 public:
-    const int EMPTY_CELL_SIGN = 0;
-    const int CUBE_SIZE = 3;
-    const std::set<int> POSSIBLE_CELL_VALUES = {1, 2, 3, 4, 5, 6, 7, 8, 9};     
-
-    SudokuBoard(int size);
-    SudokuBoard(int size, std::vector<std::vector<int>> board_values);
-    int get_size();
-    std::vector<std::vector<int>> get_board_values();
-    std::vector<std::vector<std::set<int>>> get_possible_options();
-    std::set<int> get_cell_options(int r, int c);
-    STATUS_OPTION get_status();
-    STATUS_OPTION get_cell_status(int r, int c);
-    void update_cell_value(int r, int c, int val);
-    void update_possible_options();
-
-
-    friend class TestSudokuBoard;
+  static const int EMPTY_CELL_SIGN;
+  static const size_t CUBE_SIZE;
+  static const std::set<int> POSSIBLE_CELL_VALUES;
+  static const std::string OPTION_INVALID_ERR_MSG;
+  static const std::string SIZE_INCOMPATIBALE_ERR_MSG;
+
+  // constructors
+  SudokuBoard(size_t size);
+  /**
+   * assuming that board_values is a square (its hight == its width)
+   */
+  SudokuBoard(std::vector<std::vector<int>> board_values);
+  // getters
+  size_t get_size() const;
+  const std::vector<std::vector<int>> &get_board_values() const;
+  const std::vector<std::vector<std::set<int>>> &get_possible_options() const;
+  const std::set<int> &get_cell_options(size_t r, size_t c) const;
+  size_t get_missing_cells_amount() const;
+  // setters
+  void set_cell_value(size_t r, size_t c, int val);
+  // operators
+  /**
+   * other should be in the same size as this.
+   */
+  SudokuBoard &operator=(const SudokuBoard &other);
+  // methods
+  STATUS_OPTION calc_status() const;
+  STATUS_OPTION calc_cell_status(size_t r, size_t c) const;
+  void update_possible_options();
+  std::set<int>::iterator erase_option(size_t r, size_t c,
+                                       std::set<int>::iterator option_iter);
+  bool is_empty_cell(size_t r, size_t c) const;
+
+  friend class TestSudokuBoard;
+
+private:
+  const size_t m_size;
+  std::vector<std::vector<int>> m_board_values;
+  std::vector<std::vector<std::set<int>>> m_possible_options;
+
+  // setters
+  void set_cell_options(size_t r, size_t c, std::set<int> cell_options);
+
+  // methods
+  /**
+   * calculates which values cannot appear in a specific cell, as a result of
+   * collapse with another value in the cube.
+   */
+  std::set<int> calc_forbidden_values_in_cube(size_t r, size_t c) const;
+  /**
+   * calculates which values cannot appear in a specific cell, as a result of
+   * collapse with another value in the row.
+   */
+  std::set<int> calc_forbidden_values_in_row(size_t r, size_t c) const;
```

Comment:
```suggestion
  std::set<int> calc_forbidden_values_in_row(size_t r, size_t c) const;
  
```

Add better spacings between functions